test/topology_custom/test_raft_recovery_entry_loss.py

```diff
@@ -0,0 +1,152 @@
+    - node 2: v2,
+    - nodes 3-5: v3.
+    3. Kill nodes 3-5 causing a permanent group 0 majority loss. We also permanently lose v3. After recovering majority,
+    nodes 1-2 shoud have v2, which is the only safe option.
+    4. Run the recovery procedure to recreate group 0 with nodes 1, 2 as the only members. We make sure node 2
+    becomes the leader of new group 0 since it has a newer group 0 state. When node 1 joins new group 0, it receives
+    group 0 snapshot with v2 from node 2 and applies it. If node 1 became the leader, both nodes would end up with v1.
+    5. Check that node 1 has moved its group 0 state to v2.
+    6. Remove nodes 3-5 from topology using the standard removenode procedure.
+    7. Add a new node (a sanity check verifying that the cluster is functioning properly).
+    """
+    logging.info('Adding initial servers')
+    servers = await manager.servers_add(5)
+    live_servers = servers[:2]
+    dead_servers = servers[2:]
+    logging.info(f'Servers to syrvive majority loss: {live_servers}, servers to be killed: {dead_servers}')
+
+    cql = manager.get_cql()
+
+    first_group0_id = (await cql.run_async(
+            "SELECT value FROM system.scylla_local WHERE key = 'raft_group0_id'"))[0].value
+
+    logging.info(f'Stopping {live_servers[0].server_id} to keep its group 0 state in v1')
+    await manager.server_stop(live_servers[0].server_id)
+
+    logging.info('Creating keyspace ks1, moving the group 0 state to v2')
+    await cql.run_async(
+            "CREATE KEYSPACE ks1 WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1}")
+
+    logging.info(f'Stopping {live_servers[1].server_id} to keep its group 0 state in v2')
+    await manager.server_stop(live_servers[1].server_id)
+
+    logging.info('Creating keyspace ks2, moving the group 0 state to v3')
+    await cql.run_async(
+            "CREATE KEYSPACE ks2 WITH replication = {'class': 'NetworkTopologyStrategy', 'replication_factor': 1}")
+
+    cql = await reconnect_driver(manager)
+    hosts = await wait_for_cql_and_get_hosts(cql, dead_servers, time.time() + 60)
+
+    v_group0 = await get_group0_schema_version(cql, hosts[0])
+    logging.info(f'Found group 0 schema version {v_group0}')
+
+    logging.info(f'Killing {dead_servers}')
+    for srv in dead_servers:
+        await manager.server_stop(server_id=srv.server_id)
+
+    logging.info(f'Starting {live_servers}')
+    for srv in live_servers:
+        await manager.server_start(srv.server_id)
+
+    cql = await reconnect_driver(manager)
+    hosts = await wait_for_cql_and_get_hosts(cql, live_servers, time.time() + 60)
+
+    v_node1 = await get_local_schema_version(cql, hosts[0])
+    logging.info(f'Found schema version {v_node1} on {live_servers[0].server_id}')
+    v_node2 = await get_local_schema_version(cql, hosts[1])
+    logging.info(f'Found schema version {v_node2} on {live_servers[1].server_id}')
+    assert v_group0 != v_node1 != v_node2 != v_group0
```

Comment:
won't this do something like `((a != b) != c)` so comparing `c` to the result of `(a != b)` i.e. `c != True`?

Reply:
Python `a != b != c` is functionally equivalent to `a != b and b != c` in Python. It's called something like *chaining comparison operators*.

But I think I'll change to something that won't confuse people, like `a != b and a != c and b != c`.

Reply:
> But I think I'll change it to something that won't confuse people, like `a != b and a != c and b != c`.

I did exactly that.